### ISSUE TITLE
Cache module resolution in the LSP server

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#2330]
### Change

When a module has to be resolved by the LSP server and it and all modules it depends on were previously resolved and they have not changed since, and the module was used in the previous compilation, do not resolve it again but instead use a cached result.

### Not in scope
- Resolution caching at a higher granularity than modules such as classes or methods.
- Caching parsing or verification
- Caching translation to Boogie or Boogie processing steps
- Resolving modules in parallel

### Prerequisites
- If we do https://github.com/dafny-lang/dafny/issues/2327 first, then we don't need to implement hashing for parsed modules, since we can use pointer equality.

### Enables 
- https://github.com/dafny-lang/dafny/issues/2328
- https://github.com/dafny-lang/dafny/issues/2332

### Implementation hints
- A hash of the AST of a module, and all modules it depends on, right after parsing can serve as a cache key
- There's no hash function defined for the Dafny AST, but the printed AST can serve as a hash.
- Cache entries can be removed if they were not used during a compilation

### Implementation steps
- `Resolver.ResolveProgram` has code around line 560 which clones modules in preparation for compilation. It's seems like this cloning isn't necessary and only complicates the pipeline. See if it can be removed.
- `Resolver.ResolveProgram` has several loops over modules. See if those loops can be merged. Part of the reason that there are multiple loops is that resolution can stop if there is an error in one loop. However, we could do that stopping of resolution at the module level. Ideally we extract a `ResolveModule` method that goes through the PreResolve to PostResolve phases.
- Introduce a module level granularity cache for resolution, that is used solely by the language server. Items may be pruned if they're not used for a single compilation.

**Comment by keyboardDrummer** [comment]
> Does this cover Dafny resolution or Boogie resolution? From my (limited) benchmarking, Dafny resolution isn't a large contributor to its runtime; translation and boogie-level resolution are much slower.

Good callout, made: https://github.com/dafny-lang/dafny/issues/2361

**Comment by cpitclaudel** [comment]
Excellent, thanks!
